prov/net/src/xnet_progress.c

```diff
@@ -766,14 +783,39 @@ static void xnet_progress_cqe(struct xnet_progress *progress,
 			      struct xnet_uring *uring,
 			      ofi_io_uring_cqe_t *cqe)
 {
+	struct xnet_xfer_entry *tx_entry;
 	struct ofi_sockctx *sockctx;
+	struct ofi_bsock *bsock;
+	struct xnet_ep *ep;
 
 	assert(xnet_io_uring);
 	sockctx = (struct ofi_sockctx *) cqe->user_data;
 	assert(sockctx);
 
+	bsock = (struct ofi_bsock *) sockctx->context;
+	assert(bsock);
+
+	ep = container_of(bsock, struct xnet_ep, bsock);
+
+	assert(sockctx->uring_sqe_inuse);
+	sockctx->uring_sqe_inuse = false;
 	if (&uring->ring == progress->sockapi.tx_uring.io_uring) {
+		tx_entry = ep->cur_tx.entry;
+		assert(tx_entry);
+
 		progress->sockapi.tx_uring.credits++;
+		if (cqe->res < 0) {
+			if (!OFI_SOCK_TRY_SND_RCV_AGAIN(-cqe->res))
+				xnet_complete_tx(ep, cqe->res);
+		} else {
+			ep->cur_tx.data_left -= cqe->res;
```

Comment:
Maybe assert(cqe->res >= ep->cur_tx.data_left) to ensure things haven't gotten out of whack?

Reply:
Good idea! Thanks!

Reply:
Actually, it should be `assert(cqe->res <= ep->cur_tx.data_left)`